### PR TITLE
Blog post about binplz

### PR DESCRIPTION
**Pull request by cdepillabout**

### CODE REVIEW

File: 2023-02-26-binplz.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+You can confirm that this tree binary is statically linked:
+
+```console
+$ file ./tree
+./tree: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
+```
+
+The "statically linked" part of the output of file signifies that this binary
+is statically linked.
+
```
Comment (jonascarpay) [inline 2023-02-26-binplz.md]:
```suggestion
```

I think you can safely remove it, it follows directly from the previous lines

File: 2023-02-26-binplz.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+Nixpkgs, pulls out the statically-linked `bin/tree` binary from the output, and
+returns it.
+
+`binplz.dev` should be able to build and serve any package from Nixpkgs.[^2]
+
+## Writing binplz
+
+Jonas and Hideaki joined me to build an initial version of binplz.  We created a
+[binplz repo](https://github.com/binplz/binplz.dev/) to hold our code,
+and Jonas quickly put together an initial implementation.
```
Comment (jonascarpay) [inline 2023-02-26-binplz.md:58]: I think you should mention that for a while you could totally `curl binplz.dev/tree`, it kind of sounds like we never got further than something that just runs on your own computer

File: 2023-02-26-binplz.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+The initial implementation did everything we required above, but our next task
+was getting it deployed.
```
Comment (jonascarpay) [inline 2023-02-26-binplz.md]: We _did_ deploy it, the issue was getting it to a place where it would scale

File: 2023-02-26-binplz.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+the flexibility and power of Nix and Nixpkgs.  It would be a useful service if
+you need a certain binary in a pinch.  There are a bunch of different directions
+you could take it in to do other cool things.
+
+I hope someone decides to pick up this idea and drive it to completion!
+
+## Footnotes
+
+[^1]: We were seriously considering using [nixbuild.net](https://nixbuild.net/)
+    for this.
```
Comment (jonascarpay) [inline 2023-02-26-binplz.md]: I think you should say they were very willing to help out

**Review by jonascarpay — COMMENTED** [review]
I think there's a more interesting story here than just "we ran into these three problems". The way I see it, we had something that worked, but we completely got lost in how to scale it up. Every individual problem was solvable, but the list of problems just grew so much faster than we had initially expected, or than we were able to deal with. So much so that at least I just sort of got stuck in analysis paralysis. On top of what you mentioned there were so many insidious issues like
- sorting errors into deterministic/non-deterministic, how to handle what to retry and what to fail
- managing parallel builds (memory issues, DB race conditions, hardware provisioning)
- scheduling/queuing/caching, each of which is just hard in its own right
- funding
 
 and it seemed like the longer we (or at least I) thought about the problem, the larger the distance between our PoC and something worth releasing.
 
 Anyway, I'll leave it up to you in how much detail you actually want to tell the story, I think what you have is already enough to be an interesting post. But I also feel like we/I learned a lot more than just these 3 things, so it might be valuable to talk about that to your audience.

**Comment by cdepillabout** [comment]
@jonascarpay Thanks for the suggestions.

I agree with your review, and I tried to go into a little more depth about the problems we ran into (although I didn't necessarily go into that much depth about the solutions we had come up with).

I also decided to go with your wording for some of the explanation on why we threw in the towel.

File: 2023-02-26-binplz.md (hunk 4 above)
Comment (cdepillabout) [reply 2023-02-26-binplz.md]: Good point.  I moved this from a footnote to an actual paragraph in the article in 7698efe.

File: 2023-02-26-binplz.md (hunk 1 above)
Comment (cdepillabout) [reply 2023-02-26-binplz.md]: Good idea, removed in 7698efe.

File: 2023-02-26-binplz.md (hunk 2 above)
Comment (cdepillabout) [reply 2023-02-26-binplz.md:58]: Good call, I reworded this whole section in 7698efe to say that we did at least have a good story for getting it deployed, but it was productionizing it that was the real pain.

File: 2023-02-26-binplz.md (hunk 3 above)
Comment (cdepillabout) [reply 2023-02-26-binplz.md]: Fixed in 7698efe11d814fdc94f70e46b5450c88d612eff0.

**Review by jonascarpay — APPROVED** [review]
I think this is a big improvement. My one comment would be that I think you might want to have a clearer conclusion about what you learned or want the takeaway to be for the reader.

**Comment by cdepillabout** [comment]
Thanks for the review again!

I added a little bit to the conclusion, but I guess I don't really have a clear conclusion for the project as a whole.  We ran into lots of problems that weren't necessarily hard on their own, but we just didn't come up with a solution that would be quick to implement and cheap.

I guess I don't really have a strong take-away, other than that I hope someone decides to pick up this project and run with it.  It would be a neat service to have available.

**Pull request merged**